desc/objectives/_neoclassical.py

```diff
@@ -0,0 +1,532 @@
+    get_quadrature,
+    grad_automorphism_sin,
+)
+from ..integrals.basis import FourierChebyshevSeries
+from .objective_funs import _Objective, collect_docs
+from .utils import _parse_callable_target_bounds
+
+_bounce_overwrite = {
+    "deriv_mode": """
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute Jacobian matrix, either forward mode or reverse mode AD.
+        ``auto`` selects forward or reverse mode based on the size of the input and
+        output of the objective. Has no effect on ``self.grad`` or ``self.hess`` which
+        always use reverse mode and forward over reverse mode respectively.
+
+        Default is ``fwd``. If ``rev`` is chosen, then ``jac_chunk_size=1`` is chosen
+        by default. In ``rev`` mode, reducing the pitch angle parameter ``batch_size``
+        does not reduce memory, so it is recommended to retain the default for that.
+        """
+}
+
+
+class EffectiveRipple(_Objective):
+    """The effective ripple is a proxy for neoclassical transport.
+
+    The 3D geometry of the magnetic field in stellarators produces local magnetic
+    wells that lead to bad confinement properties with enhanced radial drift,
+    especially for trapped particles. Neoclassical (thermal) transport can become the
+    dominant transport channel in stellarators which are not optimized to reduce it.
+    The effective ripple is a proxy, measuring the effective modulation amplitude of the
+    magnetic field averaged along a magnetic surface, which can be used to optimize for
+    stellarators with improved confinement. It is targeted as a flux surface function.
+
+    References
+    ----------
+    https://doi.org/10.1063/1.873749.
+    Evaluation of 1/ν neoclassical transport in stellarators.
+    V. V. Nemov, S. V. Kasilov, W. Kernbichler, M. F. Heyn.
+    Phys. Plasmas 1 December 1999; 6 (12): 4622–4632.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        ``Equilibrium`` to be optimized.
+    grid : Grid
+        Optional, tensor-product grid in (ρ, θ, ζ) with uniformly spaced nodes
+        (θ, ζ) ∈ [0, 2π) × [0, 2π/NFP). Powers of two are preferable.
+    X : int
+        Grid resolution in poloidal direction for Clebsch coordinate grid.
+        Preferably power of 2.
+    Y : int
+        Grid resolution in toroidal direction for Clebsch coordinate grid.
+        Preferably power of 2.
+    Y_B : int
+        Desired resolution for algorithm to compute bounce points.
+        Default is double ``Y``. Something like 100 is usually sufficient.
+        Currently, this is the number of knots per toroidal transit over
+        to approximate |B| with cubic splines.
+    num_transit : int
+        Number of toroidal transits to follow field line.
+        For axisymmetric devices, one poloidal transit is sufficient. Otherwise,
+        assuming the surface is not near rational, more transits will
+        approximate surface averages better, with diminishing returns.
+    num_well : int
```

Comment:
A thought:

Suppose we didn't have num_well high enough so that we were ignoring some wells. However, neoclassical optimization tends to smooth out small wells, so that after a few iterations, the actual number of wells will likely decrease, so that now num_well may be large enough? IE, could we set num_well relatively small and rely on the optimization to smooth them out successively?

Reply:
1. If you're targeting effective ripple = 0 and the optimizer succeeds yes. If you are targeting a finite ripple, you should keep in mind that in these metrics we compute some `function ~ (sum over all wells) / (length of field line)`. If num well is less than the actual number of wells over the sample of the field line, then the sum in the numerator terminates early, so the metric would be underestimated. Then the optimization may terminate early.
2. Is there a reason you suggest that over ignoring narrow wells when they are computed in `bounce.points`? That could be done fine, so that the integrals is done over `num_well` wells with `min_well_length`.
3. For ripple its clear 2 is reasonable. For Gamma_c, the pitch angle determines the contribution of the integral moreso that the width, so it's not immediately obvious to me that 2 works.
4. As you know for high NFP configurations smoothing isn't really relevant